doschema/merger.py

```diff
@@ -0,0 +1,193 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of DoSchema.
+# Copyright (C) 2016 CERN.
+#
+# DoSchema is free software; you can redistribute it
+# and/or modify it under the terms of the GNU General Public License as
+# published by the Free Software Foundation; either version 2 of the
+# License, or (at your option) any later version.
+#
+# DoSchema is distributed in the hope that it will be
+# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with DoSchema; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""Merger module."""
+
+
+class JsonSchemaCompatibilityChecker(object):
+    """Class for checking compatibility between schemas."""
+
+    def __init__(self, schema_id, curr_field, curr_schema, ignore_index=True):
+        """Constructor.
+
+        :param schema_id: Index of schema that is currently processed.
+        :param curr_field: Tuple with path to currently processed field.
+        :param curr_schema: Schema or subschema that is currently processed.
+        """
+        self.schema_id = schema_id
+        self.curr_field = curr_field
+        self.curr_schema = curr_schema
+        self.ignore_index = ignore_index
+
+    def traverse(self, fields_types_dict):
+        """Go through the schema and retrieve schema's particular fields.
+
+        :param fields_types_dict: Dictionary keeping path and type of
+                                  the field.
+        """
+        for field_name, field_value in self.curr_schema.items():
+            if field_name in ['anyOf', 'allOf', 'oneOf']:
+                self.collect_special(
+                    fields_types_dict,
+                    field_value
+                )
+            # if a value is a dict recursively comes into it
+            elif isinstance(field_value, dict):
```

Comment:
Many things can have `dict` type in a JSON Schema. Take for example [dependencies](http://spacetelescope.github.io/understanding-json-schema/reference/object.html?highlight=dependencies#dependencies). It would be better to use the fields names as these are unique in the specification.

I don't think that the current implementation works with same fields defined in different ways. It adds to the `cur_field` every key in the JSON Schema (example: `properties`) instead of just using the field names as they would appear in the JSON validated by the JSON Schemas.

Can you try this test please:
```json
{
  "type": "object",

  "properties": {
    "name": { "type": "string" },
    "credit_card": { "type": "number" }
  },

  "required": ["name"],

  "dependencies": {
    "credit_card": {
      "properties": {
        "billing_address": { "type": "string" }
      },
      "required": ["billing_address"]
    }
  }
}
```
It can be found [here](http://spacetelescope.github.io/understanding-json-schema/reference/object.html?highlight=dependencies#schema-dependencies).